include/pybind11/cast.h

```diff
@@ -231,22 +237,19 @@ class type_caster_generic {
 
         auto &internals = get_internals();
 
-        auto it = internals.registered_types_cpp.find(std::type_index(*type_info));
-        if (it == internals.registered_types_cpp.end()) {
```

Comment:
I think this one's actually the only code block where this kind of hackery is justified because it *is* the hot path, being the generic type caster. Mind if just revert this one? 

(another you didn't mention, the implicit conversions one, is also another "cold" path, likely to be just called once in the initialization, so should be cool)